### ISSUE TITLE
Low training accuracy while high ROC score

### ISSUE THREAD

**Issue by saedr** [tkipf/gae#42]
I'm using your code on a different dataset (SNOMED), so I use my own function to load the data and create all matrices similar to what you generated.

My issue is that training accuracy is decreasing throughout the training and it reaches 5% after 200 epochs while ROC and AP are increasing and are in the range of 88-90% after the training is finished.

Do you think this is an issue? i.e., are the values for ROC and AP are valid even with low training accuracy?

BTW, thanks for the awesome code!

**Comment by tkipf** [comment]
Maybe your classes are imbalanced?

**Comment by saedr** [comment]
Thanks for your prompt reply. Actually I'm doing the link prediction task. There is no notion of class in my dataset and all the other settings are similar to yours. I can only imagine something might be wrong in my data preparation. So, you think this is definitely an issue and should be addressed?

**Comment by tkipf** [comment]
What I meant is that accuracy in link prediction can be misleading if the
classes (link vs no connection) in your evaluation set are imbalanced.

**Comment by saedr** [comment]
Maybe I'm wrong but in calculating training accuracy you just use the original adjacency graph (`adj_orig`) and its reconstructed version (`model.reconstructions`):

[Line 18](https://github.com/tkipf/gae/blob/a1aecb00de894859f509f2d392135086a86946a0/gae/optimizer.py#L18) of optimizer.py:

`self.correct_prediction = tf.equal(tf.cast(tf.greater_equal(tf.sigmoid(preds_sub), 0.5), tf.int32),
                                           tf.cast(labels_sub, tf.int32))`

Please let me know if I am missing something.

**Comment by bbjy** [comment]
I meet the same problem as you @saedr , have you solve the problem? Besides, when I use the code in my own dataset, the AP score is much higher than the ROC score, I don't know why. So do you have some idea about the reason?  @saedr @tkipf 
Please let me know if your have some suggestions. Thank you so much!

**Comment by saedr** [comment]
Nope, I couldn't figure out why sometimes train_acc is decreasing while val_ap is increasing, and I don't know if this is okay or an issue!!

**Comment by tkipf** [comment]
This can be due to class imbalance when measuring accuracy.

**Comment by bbjy** [comment]
@saedr @tkipf  Thank you for your reply. I will check my data again.

**Comment by zhangxiaozao** [comment]
> Maybe I'm wrong but in calculating training accuracy you just use the original adjacency graph (`adj_orig`) and its reconstructed version (`model.reconstructions`):
> 
> [Line 18](https://github.com/tkipf/gae/blob/a1aecb00de894859f509f2d392135086a86946a0/gae/optimizer.py#L18) of optimizer.py:
> 
> `self.correct_prediction = tf.equal(tf.cast(tf.greater_equal(tf.sigmoid(preds_sub), 0.5), tf.int32), tf.cast(labels_sub, tf.int32))`
> 
> Please let me know if I am missing something.

Do you think it is not appropriate to represent the training accuracy with the original adjacency graph (adj_orig) and its reconstructed version (model.reconstructions)? I think we should just use the training edges to compute training accuracy.
